src/concurrency/weak_memory.rs

```diff
@@ -234,8 +234,11 @@ impl<'mir, 'tcx: 'mir> StoreBuffer {
     /// so a fresh atomic object has an empty store buffer and this function
     /// will return `None`. In this case, the caller should ensure that the non-buffered
     /// value from `MiriEvalContext::read_scalar()` is observed by the program, which is
-    /// the initial value of the atomic object. `MiriEvalContext::read_scalar()` is always
-    /// the latest value in modification order so it is always correct to be observed by any thread.
+    /// the initial value of the atomic object.
+    ///
+    // FIXME: This means that the initial value is no longer observable as soon as a store
+    // occurs, which limits the possible behaviours we can generate. This is an unfortunate
+    // quirk of our implementation since we can't hook into Atomic*::new().
```

Comment:
What I don't understand is why when we "create the atomic object" here (which as I understand happens lazily, on the first atomic read/write), we cannot use the current value stored there and use that to pre-fill the store buffer?

The way I think about it is like this: the weak memory emulation needs the concept of "atomic objects". That concept doesn't actually exist in the Rust language semantics, so we have to hack a bit to make this all work. We lazily create atomic objects when the first atomic access happens, and we tear down atomic objects when we notice mixed-size accesses (which could indicate that one `&mut Atomic` was destroyed and a new one at a different type was created). This can miss some weak behaviors, but at least we never throw errors for valid Rust code.

Is that a correct description of what happens? Might be good to add this to the global comment.

Reply:
It's fine to lazily initialise the store buffer on read or rmw, but I don't know what to do with the write case because we don't have the current value
https://github.com/rust-lang/miri/blob/d58abab2e74004f54ebbd778a9602b28314b9a39/src/data_race.rs#L522-L532

And I don't want to call `this.read_scalar()` because it'll end up calling `M::memory_read` which could mutate some states (such as the race detector) even though the MIR didn't actually perform any read.

Reply:
Hm, yeah we'd need to add a way to bypass the `memory_read` hook.

Is it possible to not solve this for now and leave a FIXME and document this as a limitation somewhere?